File: worker/src/main/java/com/github/k0kubun/github_ranking/worker/UserStarScanWorker.java
package com.github.k0kubun.github_ranking.worker;

import com.github.k0kubun.github_ranking.config.Config;
import com.github.k0kubun.github_ranking.github.GitHubClient;
import com.github.k0kubun.github_ranking.github.GitHubClientBuilder;
import com.github.k0kubun.github_ranking.model.User;
import com.github.k0kubun.github_ranking.repository.dao.LastUpdateDao;
import com.github.k0kubun.github_ranking.repository.dao.UserDao;
import org.skife.jdbi.v2.DBI;
import org.skife.jdbi.v2.Handle;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.sql.Timestamp;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

// Scan all starred users
public class UserStarScanWorker extends UpdateUserWorker {
    private static final long TOKEN_USER_ID = 3138447; // k0kubun
    private static final long THRESHOLD_DAYS = 1; // At least later than Mar 6th
    private static final long MIN_RATE_LIMIT_REMAINING = 500; // Limit: 5000 / h
    private static final int BATCH_SIZE = 100;
    private static final Logger LOG = LoggerFactory.getLogger(UserStarScanWorker.class);

    private final BlockingQueue<Boolean> userStarScanQueue;
    private final DBI dbi;
    private final GitHubClientBuilder clientBuilder;
    private final Timestamp updateThreshold;

    public UserStarScanWorker(Config config) {
        super(config.getDatabaseConfig().getDataSource());
        userStarScanQueue = config.getQueueConfig().getUserFullScanQueue();
        clientBuilder = new GitHubClientBuilder(config.getDatabaseConfig().getDataSource());
        dbi = new DBI(config.getDatabaseConfig().getDataSource());
        updateThreshold = Timestamp.from(Instant.now().minus(THRESHOLD_DAYS, ChronoUnit.DAYS));
    }

    @Override
    public void perform() throws Exception {
        while (userStarScanQueue.poll(5, TimeUnit.SECONDS) == null) {
            if (isStopped) {
                return;
            }
        }

        GitHubClient client = clientBuilder.buildForUser(TOKEN_USER_ID);
        LOG.info(String.format("----- started UserStarScanWorker (API: %s/5000) -----", client.getRateLimitRemaining()));
        try (Handle handle = dbi.open()) {
            // 2 * (1000 / 30 min) ≒ 4000 / hour
            int numUsers = 1000;
            while (numUsers > 0) {
                // Check rate limit
                int remaining = client.getRateLimitRemaining();
                LOG.info(String.format("API remaining: %d/5000 (numUsers: %d)", remaining, numUsers));
                if (remaining < MIN_RATE_LIMIT_REMAINING) {
                    LOG.info(String.format("API remaining is smaller than %d. Stopping.", remaining));
                    break;
                }

                // Find a current cursor
                long lastUpdatedId = handle.attach(LastUpdateDao.class).userId(LastUpdateDao.USER_STAR_SCAN);
                long stars;
                if (lastUpdatedId == 0) {
                    stars = handle.attach(UserDao.class).maxStargazersCount();
                } else {
                    stars = handle.attach(UserDao.class).userStargazersCount(lastUpdatedId);
                }

                // Query a next batch
                List<User> users = new ArrayList<>();
                while (users.isEmpty()) {
                    users = handle.attach(UserDao.class).usersWithStarsAfter(stars, lastUpdatedId, Math.min(numUsers, BATCH_SIZE));
                    if (users.isEmpty()) {
                        stars = handle.attach(UserDao.class).nextStargazersCount(stars);
                        if (stars == 0) {
                            handle.attach(LastUpdateDao.class).resetUserId(LastUpdateDao.USER_STAR_SCAN);
                            LOG.info(String.format("--- completed and reset UserStarScanWorker (API: %s/5000) ---", client.getRateLimitRemaining()));
                            return;
                        }
                        lastUpdatedId = 0;
                    }
                }

                // Update users in the batch
                LOG.info(String.format("Batch size: %d (stars: %d)", users.size(), stars));
                lastUpdatedId = updateUsers(client, handle, users, lastUpdatedId);
                handle.attach(LastUpdateDao.class).updateUserId(LastUpdateDao.USER_FULL_SCAN, lastUpdatedId);

                numUsers -= users.size();
            }
        }
        LOG.info(String.format("----- finished UserStarScanWorker (API: %s/5000) -----", client.getRateLimitRemaining()));
    }

    private long updateUsers(GitHubClient client, Handle handle, List<User> users, long lastUpdatedId) throws IOException {
        for (User user : users) {
            Timestamp updatedAt = handle.attach(UserDao.class).userUpdatedAt(user.getId()); // TODO: Fix N+1
            if (updatedAt.before(updateThreshold)) {
                updateUser(handle, user, client);
                LOG.info(String.format("[%s] userId = %d (stars: %d)", user.getLogin(), user.getId(), user.getStargazersCount()));
            } else {
                LOG.info(String.format("Skip up-to-date user (id: %d, login: %s, updatedAt: %s)", user.getId(), user.getLogin(), updatedAt.toString()));
            }

            if (lastUpdatedId < user.getId()) {
                lastUpdatedId = user.getId();
            }
            if (isStopped) { // Shutdown immediately if requested
                break;
            }
        }
        return lastUpdatedId;
    }
}

File: worker/src/main/kotlin/com/github/k0kubun/gitstar_ranking/workers/UserStarScanWorker.kt
package com.github.k0kubun.gitstar_ranking.workers

import com.github.k0kubun.gitstar_ranking.GitstarRankingConfiguration
import com.github.k0kubun.gitstar_ranking.client.GitHubClient
import com.github.k0kubun.gitstar_ranking.client.GitHubClientBuilder
import com.github.k0kubun.gitstar_ranking.core.User
import com.github.k0kubun.gitstar_ranking.db.LastUpdateQuery
import com.github.k0kubun.gitstar_ranking.db.STAR_SCAN_STARS
import com.github.k0kubun.gitstar_ranking.db.STAR_SCAN_USER_ID
import com.github.k0kubun.gitstar_ranking.db.UserQuery
import java.sql.Timestamp
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.concurrent.BlockingQueue
import java.util.concurrent.TimeUnit
import org.jooq.impl.DSL.using
import org.slf4j.LoggerFactory

val PENDING_USERS = listOf(
    // Users with too many repositories. To be fixed later.
    "GITenberg",
    "gitpan",
    "the-domains",
    "wp-plugins",
    "gitter-badger",
    // Somehow 502?
    "Try-Git",
)

private const val TOKEN_USER_ID: Long = 3138447 // k0kubun
private const val THRESHOLD_DAYS: Long = 7 // At least later than Mar 6th
private const val MIN_RATE_LIMIT_REMAINING: Long = 500 // Limit: 5000 / h
private const val BATCH_SIZE = 100

// Scan all starred users
class UserStarScanWorker(config: GitstarRankingConfiguration) : UpdateUserWorker(config.database.dslContext) {
    private val logger = LoggerFactory.getLogger(UserStarScanWorker::class.simpleName)
    private val userStarScanQueue: BlockingQueue<Boolean> = config.queue.userStarScanQueue
    private val database = config.database.dslContext
    private val clientBuilder: GitHubClientBuilder = GitHubClientBuilder(config.database.dslContext)
    private val updateThreshold: Timestamp = Timestamp.from(Instant.now().minus(THRESHOLD_DAYS, ChronoUnit.DAYS))

    override fun perform() {
        while (userStarScanQueue.poll(5, TimeUnit.SECONDS) == null) {
            if (isStopped) {
                return
            }
        }
        val client = clientBuilder.buildForUser(TOKEN_USER_ID)
        logger.info("----- started UserStarScanWorker (API: ${client.rateLimitRemaining}/5000) -----")
        var numUsers = 1000 // 2 * (1000 / 30 min) ≒ 4000 / hour
        var numChecks = 2000 // Avoid issuing too many queries by skips
        while (numUsers > 0 && numChecks > 0 && !isStopped) {
            // Find a current cursor
            var lastUpdatedId = LastUpdateQuery(database).findCursor(key = STAR_SCAN_USER_ID) ?: 0L
            var stars = LastUpdateQuery(database).findCursor(key = STAR_SCAN_STARS) ?: 0L
            if (stars == 0L) {
                stars = UserQuery(database).max("stargazers_count") ?: 0L
            }

            // Query a next batch
            var users = emptyList<User>()
            while (users.isEmpty()) {
                users = UserQuery(database).orderByIdAsc(
                    stargazersCount = stars, idAfter = lastUpdatedId, limit = numUsers.coerceAtMost(BATCH_SIZE),
                )
                if (users.isEmpty()) {
                    stars = UserQuery(database).findStargazersCount(stargazersCountLessThan = stars) ?: 0L
                    if (stars == 0L) {
                        LastUpdateQuery(database).delete(key = listOf(STAR_SCAN_USER_ID, STAR_SCAN_STARS))
                        logger.info("--- completed and reset UserStarScanWorker (API: ${client.rateLimitRemaining}/5000) ---")
                        return
                    }
                    lastUpdatedId = 0
                }
            }

            // Update users in the batch
            logger.info("Batch size: ${users.size} (stars: $stars)")
            for (user in users) {
                if (PENDING_USERS.contains(user.login)) {
                    logger.info("Skipping a user with too many repositories: ${user.login}")
                    continue
                }

                // Check rate limit
                val remaining = client.rateLimitRemaining
                logger.info("API remaining: $remaining/5000 (numUsers: $numUsers, numChecks: $numChecks)")
                if (remaining < MIN_RATE_LIMIT_REMAINING) {
                    logger.info("API remaining is smaller than $remaining. Stopping.")
                    numChecks = 0
                    break
                }
                val updatedAt = UserQuery(database).find(id = user.id)!!.updatedAt // TODO: Fix N+1
                if (updatedAt.before(updateThreshold)) {
                    updateUser(user, client, TOKEN_USER_ID)
                    logger.info("[${user.login}] userId = ${user.id} (stars: ${user.stargazersCount})")
                    numUsers--
                } else {
                    logger.info("Skip up-to-date user (id: ${user.id}, login: ${user.login}, updatedAt: ${updatedAt})")
                }
                numChecks--
                if (lastUpdatedId < user.id) {
                    lastUpdatedId = user.id
                }
                if (isStopped) { // Shutdown immediately if requested
                    break
                }
            }

            // Update the counter
            val nextUpdatedId = lastUpdatedId
            val nextStars = stars
            database.transaction { tx ->
                LastUpdateQuery(using(tx)).update(key = STAR_SCAN_USER_ID, cursor = nextUpdatedId)
                LastUpdateQuery(using(tx)).update(key = STAR_SCAN_STARS, cursor = nextStars)
            }
        }
        logger.info("----- finished UserStarScanWorker (API: ${client.rateLimitRemaining}/5000) -----")
    }

    override fun updateUser(user: User, client: GitHubClient, tokenUserId: Long) {
        super.updateUser(user, client, tokenUserId)
        Thread.sleep(500) // 0.5s: 1000 * 0.5s = 500s = 8.3 min (out of 15 min)
    }
}
